lib/extensions/ar_miq_set.rb

```diff
@@ -92,8 +106,37 @@ def model_table_name
     end
   end
 
-  def add_member(member)
-    raise "object of type #{member.class} may not be a member of a set of type #{self.class}" unless member.kind_of?(self.class.model_class)
-    with_relationship_type("membership") { add_child(member) }
+  def members=(*members)
+    transaction do
+      remove_all_members
+      add_members(*members)
+    end
+  end
+  alias replace_children members=
+
+  def remove_member(member)
+    miq_set_memberships.where(:member => member).delete_all
+  end
+  alias remove_child remove_member
+
+  def remove_all_members
+    miq_set_memberships.delete_all
+  end
+  alias remove_all_children remove_all_members
+
+  def add_members(*members)
+    transaction do
+      existing = miq_set_memberships.index_by { |ms| [ms.member_type, ms.member_id] }
+
+      members.flat_map { |m| Array(m) }.each do |member|
+        raise ArgumentError, "object of type #{member.class} may not be a member of a set of type #{self.class}" unless member.kind_of?(self.class.model_class)
+        next if existing.include?([member.class.base_class.name, member.id])
+
+        miq_set_memberships.create(:member => member)
```

Comment:
Should this do `create!` ?  It's pretty benign, but I think failure should fail the whole transaction.

Reply:
See above

Reply:
Ah ok, yeah when I wrote the PoC branch, I just wrote (I don't even think I tested it fully).

So, I think we should update this to `.create!` in this PR.